nvflare/private/fed/client/client_runner.py

```diff
@@ -337,7 +346,13 @@ def abort(self):
         """
         self._abort_current_task()
         with self.engine.new_context() as fl_ctx:
-            self.fire_event(EventType.END_RUN, fl_ctx)
+            with self.end_run_lock:
+                if not self.end_run_fired:
+                    self.fire_event(EventType.ABOUT_TO_END_RUN, fl_ctx)
+                    self.log_info(fl_ctx, "ABOUT_TO_END_RUN fired")
+                    self.fire_event(EventType.END_RUN, fl_ctx)
+                    self.log_info(fl_ctx, "END_RUN fired")
+                    self.end_run_fired = True
```

Comment:
We still need to trigger self.run_abort_signal here; otherwise the try_run() won't break from the loop.